File: src/main/java/kitchenpos/order/eatinorders/application/EatInOrderService.java
package kitchenpos.order.eatinorders.application;

import kitchenpos.order.domain.Order;
import kitchenpos.order.domain.OrderRepository;
import kitchenpos.order.eatinorders.domain.EatInOrderAcceptService;
import kitchenpos.order.eatinorders.domain.EatInOrderCompleteService;
import kitchenpos.order.eatinorders.domain.EatInOrderCreateService;
import kitchenpos.order.eatinorders.domain.EatInOrderServeService;
import org.springframework.stereotype.Component;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.NoSuchElementException;
import java.util.UUID;

@Component
public class EatInOrderService {

    private final OrderRepository orderRepository;
    private final EatInOrderCreateService eatInOrderCreateService;
    private final EatInOrderAcceptService eatInOrderAcceptService;
    private final EatInOrderServeService eatInOrderServeService;
    private final EatInOrderCompleteService eatInOrderCompleteService;

    public EatInOrderService(OrderRepository orderRepository, EatInOrderCreateService eatInOrderCreateService, EatInOrderAcceptService eatInOrderAcceptService, EatInOrderServeService eatInOrderServeService, EatInOrderCompleteService eatInOrderCompleteService) {
        this.orderRepository = orderRepository;
        this.eatInOrderCreateService = eatInOrderCreateService;
        this.eatInOrderAcceptService = eatInOrderAcceptService;
        this.eatInOrderServeService = eatInOrderServeService;
        this.eatInOrderCompleteService = eatInOrderCompleteService;
    }

    @Transactional(readOnly = false)
    public Order create(Order request) {
        return this.eatInOrderCreateService.create(request);
    }

    @Transactional(readOnly = false)
    public Order accept(final UUID orderId) {
        Order order = getOrder(orderId);
        return this.eatInOrderAcceptService.accept(order);
    }

    @Transactional(readOnly = false)
    public Order serve(final UUID orderId) {
        Order order = getOrder(orderId);
        return this.eatInOrderServeService.serve(order);
    }

    @Transactional(readOnly = false)
    public Order complete(final UUID orderId) {
        Order order = getOrder(orderId);
        return eatInOrderCompleteService.complete(order);
    }

    @Transactional(readOnly = true)
    public Order getOrder(final UUID orderId) {
        return this.orderRepository.findById(orderId)
                .orElseThrow(NoSuchElementException::new);
    }

    @Transactional(readOnly = true)
    public List<Order> findAll() {
        return orderRepository.findAll();
    }
}

File: src/main/kotlin/kitchenpos/tobe/eatinorder/application/EatInOrderService.kt
package kitchenpos.tobe.eatinorder.application

import kitchenpos.tobe.eatinorder.application.dto.request.CreateOrderRequest
import kitchenpos.tobe.eatinorder.application.dto.response.CreateEatInOrderResponse
import kitchenpos.tobe.eatinorder.application.dto.response.GetOrderResponse
import kitchenpos.tobe.eatinorder.domain.EatInOrderCompleteService
import kitchenpos.tobe.eatinorder.domain.MenuClient
import kitchenpos.tobe.eatinorder.domain.entity.EatInOrder
import kitchenpos.tobe.eatinorder.domain.entity.OrderLineItemV2
import kitchenpos.tobe.eatinorder.domain.entity.OrderTableV2
import kitchenpos.tobe.eatinorder.domain.repository.EatInOrderRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDateTime
import java.util.*

@Service
@Transactional
class EatInOrderService(
    private val menuClient: MenuClient,
    private val eatInOrderRepository: EatInOrderRepository,
    private val eatInOrderCompleteService: EatInOrderCompleteService,
) {
    fun createOrder(request: CreateOrderRequest): CreateEatInOrderResponse {
        val menus = menuClient.getMenus(request.orderLineItems.map { it.menuId })
        require(menus.all { it.displayed }) { "미노출 메뉴가 존재합니다." }

        val order =
            EatInOrder.of(
                orderDateTime = LocalDateTime.now(),
                orderLineItems =
                    request.orderLineItems.map { orderLineItem ->
                        OrderLineItemV2(
                            menuId = orderLineItem.menuId,
                            quantity = orderLineItem.quantity,
                            price = orderLineItem.price,
                        )
                    },
                orderTable =
                    OrderTableV2(
                        id = request.orderTable.id,
                        name = request.orderTable.name,
                        numberOfGuests = request.orderTable.numberOfGuests,
                    ),
            )

        return CreateEatInOrderResponse.from(eatInOrderRepository.save(order))
    }

    fun accept(orderId: UUID): UUID {
        val eatInOrder = eatInOrderRepository.findEatInOrderById(orderId)
        eatInOrder.accept()
        return eatInOrder.id
    }

    fun serve(orderId: UUID): UUID {
        val eatInOrder = eatInOrderRepository.findEatInOrderById(orderId)
        eatInOrder.serve()
        return eatInOrder.id
    }

    fun complete(orderId: UUID): UUID {
        return eatInOrderCompleteService.complete(orderId)
    }

    fun findAll(): List<GetOrderResponse> {
        return eatInOrderRepository.findAll().map { GetOrderResponse.from(it) }
    }
}
